rosbag2_transport/test/rosbag2_transport/test_record_services.cpp

```diff
@@ -0,0 +1,147 @@
+      std::move(writer_), storage_options_, record_options, recorder_name_);
+    recorder_->record();
+
+    auto string_message = get_messages_strings()[1];
+    rosbag2_test_common::PublicationManager pub_manager;
+    pub_manager.setup_publisher(test_topic_, string_message, 50);
+
+    const std::string ns = "/" + recorder_name_;
+    cli_snapshot_ = client_node_->create_client<Snapshot>(ns + "/snapshot");
+    topic_sub_ = client_node_->create_subscription<test_msgs::msg::Strings>(
+      test_topic_, 10,
+      std::bind(&RecordSrvsTest::subscription_callback, this, std::placeholders::_1));
+
+    exec_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
+
+    exec_->add_node(recorder_);
+    exec_->add_node(client_node_);
+    spin_thread_ = std::thread(
+      [this]() {
+        exec_->spin();
+      });
+
+    ASSERT_TRUE(pub_manager.wait_for_matched(test_topic_.c_str()));
+    pub_manager.run_publishers();
+
+    // Make sure all expected services are present before starting any test
+    ASSERT_TRUE(cli_snapshot_->wait_for_service(service_wait_timeout_));
+  }
+
+  /// Call a service client, and expect it to successfully return within a reasonable timeout
+  template<typename Srv>
+  typename Srv::Response::SharedPtr successful_call(
+    typename rclcpp::Client<Srv>::SharedPtr cli,
+    typename Srv::Request::SharedPtr request)
+  {
+    auto future = cli->async_send_request(request);
+    EXPECT_EQ(future.wait_for(service_call_timeout_), std::future_status::ready);
+    EXPECT_TRUE(future.valid());
+    auto result = std::make_shared<typename Srv::Response>();
+    EXPECT_NO_THROW({result = future.get();});
+    EXPECT_TRUE(result);
+    return result;
+  }
+
+  template<typename Srv>
+  typename Srv::Response::SharedPtr successful_call(typename rclcpp::Client<Srv>::SharedPtr cli)
+  {
+    auto request = std::make_shared<typename Srv::Request>();
+    return successful_call<Srv>(cli, request);
+  }
+
+public:
+  // Basic configuration
+  const std::string recorder_name_ = "rosbag2_recorder_for_test_srvs";
+  const std::chrono::seconds service_wait_timeout_ {2};
+  const std::chrono::seconds service_call_timeout_ {1};
+  const std::string test_topic_ = "/recorder_srvs_test_topic";
+
+  // Orchestration
+  std::thread spin_thread_;
+  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec_;
+  std::shared_ptr<rosbag2_transport::Recorder> recorder_;
+
+  // Service clients
+  rclcpp::Node::SharedPtr client_node_;
+  rclcpp::Client<Snapshot>::SharedPtr cli_snapshot_;
+  rclcpp::Subscription<test_msgs::msg::Strings>::SharedPtr topic_sub_;
+};
+
+TEST_F(RecordSrvsTest, trigger_snapshot)
+{
+  auto & writer = recorder_->get_writer_handle();
+  MockSequentialWriter & mock_writer =
+    static_cast<MockSequentialWriter &>(writer.get_implementation_handle());
+  EXPECT_THAT(mock_writer.get_messages().size(), Eq(0u));
+  successful_call<Snapshot>(cli_snapshot_);
```

Comment:
I would recommend to add extra check with wait for some timeout that `mock_writer.snapshot_buffer_.size() > 0` before calling `successful_call<Snapshot>(cli_snapshot_);`
I am afraid that it could be a race condition when we returned from sending service request for `snapshot` while no one test messages has been recorded yet. It could potentially cause flakiness in test.